File: leetcode/golang/1456_maximum_number_of_vowels.go
func max(a int, b int) int {
    if a > b{
        return a
    }
    return b
}

func maxVowels(s string, k int) int {
    vowels := map[byte]bool {
        'a':true,
        'e':true,
        'i':true,
        'o':true,
        'u':true,
    }
    res := 0
    for l, r , tmp :=0, 0, 0; r < len(s); r++ {
        if vowels[s[r]] {
            tmp++
        }
        if r >= l + k {
            if vowels[s[l]] {
                tmp--
            }
            l++
        }
        res = max(res, tmp)
    }
    return res
}

File: leetcode/c++/1456_maximum_number_of_vowels.cpp
// https://leetcode.cn/problems/maximum-number-of-vowels-in-a-substring-of-given-length/description/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxVowels (string s, int k) {
        unordered_set<char> st = {'a', 'e', 'i', 'o', 'u'};
        int res = 0; 
        for (int l = 0, r = 0, tmp = 0; r < s.length(); r++) {
            if (st.count(s[r]))
                tmp++;
            if (r >= l + k) {
                if (st.count(s[l]))
                    tmp--;
                l++;
            }
            res = max(res, tmp);
        }
        return res;
    }
};

// 时间复杂度为O(n)，空间复杂度为O(1)